src/components/KCodeBlock/KCodeBlock.vue

```diff
@@ -0,0 +1,968 @@
+})
+
+watch(() => isRegExpMode.value, function() {
+  // Updates the matching line numbers because the matches can be different for the same query between normal and regexp mode.
+  updateMatchingLineNumbers()
+})
+
+watch(() => isShowingFilteredCode.value, async function() {
+  // Moves the focus to the code block so that code block-scoped shortcuts still work. That’s necessary because toggling filter mode changes which pre element is rendered. In doing so, the currently focused element is removed from the DOM and in response, the browser moves the focus to document.body.
+  if (document?.activeElement?.tagName === 'PRE') {
+    codeBlock.value?.focus({ preventScroll: true })
+  }
+
+  if (!isShowingFilteredCode.value) {
+    // Waits one Vue tick in which the full code block is re-rendered after switching off filter mode. Only then does it make sense to emit the corresponding event.
+    await nextTick()
+
+    // Turning off filter mode causes the full code block to be re-rendered
+    emitCodeBlockRenderEvent()
+    updateMatchingLineNumbers()
+  }
+})
+
+onMounted(function() {
+  document.addEventListener('keydown', triggerShortcuts)
+
+  if (codeBlockSearchInput.value instanceof HTMLInputElement && props.query !== '') {
+    codeBlockSearchInput.value.value = props.query
+  }
+
+  emitCodeBlockRenderEvent()
+  updateMatchingLineNumbers()
+})
+
+onBeforeUnmount(function() {
+  document.removeEventListener('keydown', triggerShortcuts)
+})
+
+function emitCodeBlockRenderEvent(): void {
+  const preElement = codeBlock.value?.querySelector('.k-highlighted-code-block')
+  const codeElement = preElement?.querySelector('code')
+
+  if (preElement instanceof HTMLElement && codeElement instanceof HTMLElement) {
+    emit('code-block-render', getEventData(preElement, codeElement))
+  }
+}
+
+function emitMatchingLinesChangeEvent(): void {
+  const preElement = codeBlock.value?.querySelector('.k-highlighted-code-block')
+  const codeElement = preElement?.querySelector('code')
+
+  if (preElement instanceof HTMLElement && codeElement instanceof HTMLElement) {
+    emit('matching-lines-change', getEventData(preElement, codeElement))
+  }
+}
+
+function getEventData(preElement: HTMLElement, codeElement: HTMLElement): CodeBlockEventData {
+  return {
+    preElement,
+    codeElement,
+    code: props.code,
+    language: props.language,
+    query: query.value,
+    matchingLineNumbers: matchingLineNumbers.value,
+  }
+}
+
+function handleSearch(event: Event): void {
+  const input = event.target as HTMLInputElement
+
+  // Ensures that no wasted debouncing takes place when the search input value being emitted is the same as the previous query. This also avoids unnecessarily showing the processing icon.
+  if (input.value !== query.value) {
+    isProcessingInternally.value = true
+    debouncedHandleSearchInputValue(input.value)
+  }
+}
+
+function handleSearchInputValue(inputValue: string): void {
+  query.value = inputValue
```

Comment:
question: Would it make sense to `trim` the search string here?

Reply:
In other search contexts, I would’ve considered trimming the input, but since this will often involve code search, I rather prioritize the flexibility of searching for string with leading and/or trailing space because there are actually use cases for it (e.g. spotting a malformed configuration by searching for `": " value` where the space before `value` is erroneous).